### ISSUE TITLE
Support onScroll and RefreshControl props

### ISSUE THREAD

**Issue by rickchi2003** [computerjazz/react-native-draggable-flatlist#135]
1. When we passed in onScroll props, it doesn't work and didn't get called
2. RefreshControl didn't work as well on Android, when adding RefreshControl, it will block the whole FlatList from scrolling.

**Comment by curiousdustin** [comment]
See #127 for more info about onScroll.

I haven't tried RefreshControl yet.

**Comment by migue1s** [comment]
Confirmed this is happening on the `refreshControl`, having it be a prop will block scrolling on Android.

**Comment by Peretz30** [comment]
In 2.3.1 you can use **onScrollOffsetChange** prop instead of onScroll. But still there is nothing about **refreshControll**

**Comment by ruslanaplid** [comment]
refreshControl not working (with props onRefresh and refreshing)

**Comment by Peretz30** [comment]
@curiousdustin any updates about `RefreshControl`?

**Comment by curiousdustin** [comment]
No, sorry, it is not something I needed on my project, so I have never investigated.

**Comment by Peretz30** [comment]
@computerjazz any plans about `RefreshControl` support?

**Comment by computerjazz** [comment]
The root issue seems to be that the `PanGestureHandler` that wraps the flatlist has a bad interaction with `RefreshControl`. I spent a while trying to find a fix but didn't find one. 

At this point I'm leaning toward a custom `RefreshControl` implementation would be the way to go, as it would keep everything in the RNGH + Reanimated world.

**Comment by Peretz30** [comment]
@computerjazz thank you for reply. I created custom `RefreshControl` Android implementation for my project. If someone need it too. I could make a PR.

**Comment by mpaschenko** [comment]
> @computerjazz thank you for reply. I created custom `RefreshControl` Android implementation for my project. If someone need it too. I could make a PR.

please make a PR, I got need of that as well, thanks!

**Comment by Peretz30** [comment]
@mpaschenko created #175

**Comment by bradbumbalough** [comment]
Any updates on this? I'm seeing that on Android there is no scrolling at all when a RefreshControl is added.

**Comment by callaars** [comment]
You can disable the refresh controller to overcome this issue.

```
<RefreshControl enabled={!isCurrentlyDragging} ... />
```

**Comment by hakanpinar** [comment]
I'm facing this issue in Android. Will there be a fix? 

RN version: 0.63.2
react-native-draggable-flatlist version: 2.4.0

**Comment by callaars** [comment]
I think there will never be a fix as the issue is with the gesture handler by another library. What I did to overcome this is to put the refresh control on a higher level than the draggable flat list. That way it works just fine on Android.

**Comment by pandeyrasu** [comment]
> I think there will never be a fix as the issue is with the gesture handler by another library. What I did to overcome this is to put the refresh control on a higher level than the draggable flat list. That way it works just fine on Android.
 Can you post your solution here?

**Comment by dev-sathish** [comment]
> I think there will never be a fix as the issue is with the gesture handler by another library. What I did to overcome this is to put the refresh control on a higher level than the draggable flat list. That way it works just fine on Android.

Please give us some example which will be very useful. 
Thanks in Advance..!

**Comment by callaars** [comment]
I will in the weekend, sorry for the delay.

**Comment by Lukkub** [comment]
> I think there will never be a fix as the issue is with the gesture handler by another library. What I did to overcome this is to put the refresh control on a higher level than the draggable flat list. That way it works just fine on Android.

@callaars Could you share your wrapper idea ?

**Comment by callaars** [comment]
I'm sorry I will do today!

**Comment by Lukkub** [comment]
@callaars 
Sorry for bothering you, but I just stuck with that issue - that will be super helpful if you could share your idea :)

**Comment by Lukkub** [comment]
@callaars

**Comment by callaars** [comment]
Ok I'm very sorry it kept going through the cracks. I've put the basics in [this gist](https://gist.github.com/callaars/4ded1e912004c264dace70e3cf44bd8a). I haven't ran this but that's basically of the components you need.

Ignore the sloppy code, as I've copy pasted things from the source tree, so some ways of implementation might not make total sense.

The main idea is that you have an overriding `ScrollView` on a higher level than the draggable flatlist. The only caveat here is that the refresh control itself doesn't work **on** the list, but on the wrapping component of it. For us, that works just fine, but if it doesn't for you then I'm sorry and I don't have a good solution for it. For us triggering the dragging was a major issue as well as it would trigger the refresh control, hence the drag aware context as well.

The meat is found in `some-component.tsx`. Again, sorry for the messy code, if you have any questions let me know.

**Comment by Lukkub** [comment]
@callaars Thanks a lot - I will jump into the code and let you know about the progress :)

**Comment by nishanttatva** [comment]
@callaars Thanks for that gist.
But are you facing any issues with scrolling when wrapping the DraggableFlatList with ScrollView? I can't seem to find a fix for that.

**Comment by AbdelkhalekESI** [comment]
Any solutions for this? The issue is still happening on android. works fine on iOS.

**Comment by AbdelkhalekESI** [comment]
@Peretz30 Tried your solution, doesn't work either on android.

**Comment by jwfwessels** [comment]
@computerjazz 
> The root issue seems to be that the `PanGestureHandler` that wraps the FlatList has a bad interaction with `RefreshControl`. I spent a while trying to find a fix but didn't find one.
> 
> At this point I'm leaning toward a custom `RefreshControl` implementation would be the way to go, as it would keep everything in the RNGH + Reanimated world.

I seem to have managed to get it working on android with a minor library patch.

I noticed something interesting, it would be good if someone could confirm and I am curious what the trade-off is of the change. Switching the dependency for `FlatList` from `react-native-gesture-handler` to `react-native`'s implementation seems to resolve the issues around `refreshControl` on android. I also needed to set the `activationDistance` prop. It seems only to be effective over `14` (`activationDistance={14}`) not sure why that is. Didn't dive too deeply into the code. 

If anyone could check if this works for them or if the maintainers could comment, I would appreciate the feedback.

**Comment by scaralfred** [comment]
Has anyone manage to fix this?

**Comment by jo-rdan** [comment]
If you are importing `ScrollView` from `react-native-gesture-handler` it won't work unless you import it from `react-native`. This solved my issue, I hope it does help you as well

**Comment by scaralfred** [comment]
Hi @jo-rdan thanks for the feedback. Exactly in which file and line of code I should replace that?

**Comment by svenlombaert** [comment]
Can confirm that using FlatList under the hood in combination with the enable/.disable of the RefreshControl works great.

I think that this PR would give us the ability to solve the problem?
https://github.com/computerjazz/react-native-draggable-flatlist/pull/288

**Comment by scaralfred** [comment]
I can confirm too, it works in that way. Sorry for the previous comments.

**Comment by gdoudeng** [comment]
android pull to refresh not work

**Comment by minhchinh97h** [comment]
I was able to have `refreshControl` prop working in Android by creating a native wrapper for `RefreshControl`, forcing the `DraggableFlatList`'s underlaid `FlatList` to prioritise the native `RefreshControl` and controlling the `RefreshControl`'s `enabled` prop on with `onDragBegin` & `onDragEnd` handlers to prevent refresh-on-dragging issue (while at the very top of the list). Here's my working version:

````
import React from 'react'
import {RefreshControl} from 'react-native'
import {createNativeWrapper} from 'react-native-gesture-handler'
import DraggableFlatList from 'react-native-draggable-flatlist'

const AndroidRefreshControl = createNativeWrapper(RefreshControl, {
  disallowInterruption: true,
  shouldCancelWhenOutside: false
})

export class MyDraggableList extends React.Component{
 refreshControlRef = React.createRef()

 state = {
  enabled: true
 }

 render(){
   <DraggableFlatList 
     ... 
     onDragBegin={() => {this.setState({enabled: false})}} 
     onDragEnd={() => {this.setState({enabled: true})}}
     waitFor={this.refreshControlRef}
     refreshControl={
      <AndroidRefreshControl
        ...
        ref={this.refreshControlRef}
        enabled={this.state.enabled}
      />
     }
   />
 }
}
````

The original idea of the workaround was from https://github.com/software-mansion/react-native-gesture-handler/issues/1067#issuecomment-898400208.

**Comment by thanhluantl2304** [comment]
any update on this? no solution worked for me! @@

**Comment by mordechaim** [comment]
❗❗ 

To fix this issue, use the `RefreshControl` component from `react-native-gesture-handler` instead:

```diff
-import { RefreshControl } from 'react-native'
+import { RefreshControl } from 'react-native-gesture-handler'
```

**Comment by AdvMaple** [comment]
> exclamationexclamation
> 
> To fix this issue, use the `RefreshControl` component from `react-native-gesture-handler` instead:
> 
> ```diff
> -import { RefreshControl } from 'react-native'
> +import { RefreshControl } from 'react-native-gesture-handler'
> ```
 This work for me
